src/platforms/windows/daemon/windowsfirewall.cpp

```diff
@@ -0,0 +1,874 @@
+  DWORD    result = ERROR_SUCCESS;
+  // Use dynamic sessions for efficiency and safety:
+  //  -> Filtering policy objects are deleted even when the application crashes/ deamon goes down
+  FWPM_SESSION0 session;
+  memset(&session, 0, sizeof(session));
+  session.flags = FWPM_SESSION_FLAG_DYNAMIC;
+
+  logger.log() << ("Opening the filter engine.");
+
+  result = FwpmEngineOpen0(
+      NULL,
+      RPC_C_AUTHN_WINNT,
+      NULL,
+      &session,
+      &engineHandle );
+
+  if (result != ERROR_SUCCESS){
+    WindowsCommons::windowsLog("FwpmEngineOpen0 failed");
+    return;
+  }
+  logger.log() << ("Filter engine opened successfully.");
+  m_sessionHandle = engineHandle;
+
+
+}
+
+WindowsFirewall::~WindowsFirewall() {
+  MVPN_COUNT_DTOR(WindowsFirewall);
+  CloseHandle(m_sessionHandle);
+}
+
+bool WindowsFirewall::init(){
+    if(m_init){
+      logger.log() << "Alread initialised FW_WFP layer";
+      return true;
+    }
+    if(m_sessionHandle == INVALID_HANDLE_VALUE){
+        logger.log() << "Cant Init Sublayer with invalid wfp handle";
+        return false;
+    }
+    // If we were not able to aquire a handle, this will fail anyway.
+    // We need to open up another handle because of wfp rules:
+    // If a wfp resource was created with SESSION_DYNAMIC,
+    // the session exlusively owns the resource, meaning the driver can't add filters to
+    // the sublayer.
+    // So let's have non dynamic session only for the sublayer creation.
+    // This means the Layer exists until the next Reboot.
+    DWORD result = ERROR_SUCCESS;
+    HANDLE wfp = INVALID_HANDLE_VALUE;
+    FWPM_SESSION0 session;
+    memset(&session, 0, sizeof(session));
+
+    logger.log() << "Opening the filter engine";
+    result = FwpmEngineOpen0(
+        NULL,
+        RPC_C_AUTHN_WINNT,
+        NULL,
+        &session,
+        &wfp );
+    if (result != ERROR_SUCCESS){
+        logger.log() << "FwpmEngineOpen0 failed. Return value:.\n" << result;
+        return false;
+    }
+    auto cleanup = qScopeGuard([&] {
+         FwpmEngineClose0(wfp);
+     });
+
+    // Check if the Layer Already Exists
+    FWPM_SUBLAYER0* maybeLayer;
+    result = FwpmSubLayerGetByKey0(wfp,&ST_FW_WINFW_BASELINE_SUBLAYER_KEY,&maybeLayer);
+    if(result == ERROR_SUCCESS){
+        logger.log() << "The Sublayer Already Exists!";
+        FwpmFreeMemory0((void**) &maybeLayer);
+        return true;
+    }
+
+    //Step 1: Start Transaction
+    result = FwpmTransactionBegin(wfp,NULL);
+    if(result != ERROR_SUCCESS){
+        logger.log() << "FwpmTransactionBegin0 failed. Return value:.\n" << result;
+        return false;
+    }
+
+    // Step 3: Add Sublayer
+    FWPM_SUBLAYER0 subLayer;
+    memset(&subLayer, 0, sizeof(subLayer));
+    subLayer.subLayerKey = ST_FW_WINFW_BASELINE_SUBLAYER_KEY;
+    subLayer.displayData.name = (PWSTR)L"MozillaVPN-SplitTunnel-Sublayer";
+    subLayer.displayData.description = (PWSTR)L"Filters that enforce a good baseline";
```

Comment:
Is this string visible somewhere? Do we need to localize? probably not.

Reply:
Not visible, this is just for internal tools :)